File: app/src/main/java/com/monk/customview/MaxHeightScrollView.java
package com.monk.customview;

import android.app.Activity;
import android.content.Context;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.view.Display;
import android.widget.ScrollView;

import com.monk.utils.LogUtil;

/**
 * @author monk
 * @date 2019-01-25
 */
public class MaxHeightScrollView extends ScrollView {
    private final String tag = "MaxHeightScrollView";

    public MaxHeightScrollView(Context context) {
        this(context, null);
    }

    public MaxHeightScrollView(Context context, AttributeSet attrs) {
        super(context, attrs);
        // 方式1
        DisplayMetrics dm = context.getResources().getDisplayMetrics();

        String sb = "heightPixels:" + dm.heightPixels + "\t" +
                "widthPixels:" +  dm.widthPixels + "\n" +
                "xDpi:" + dm.xdpi + "\t" +
                "yDpi:" + dm.ydpi + "\n" +
                "density:" + dm.density + "\t" +
                "densityDpi:" + dm.densityDpi + "\t";
        LogUtil.i(tag, sb);
        LogUtil.v(tag, "toString():" + dm.toString());

        // 方式2
        Display display = ((Activity) context).getWindowManager().getDefaultDisplay();
        DisplayMetrics displayMetrics1 = new DisplayMetrics();
        display.getMetrics(displayMetrics1);
        LogUtil.v(tag, "toString():" + displayMetrics1.toString());
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        heightMeasureSpec = MeasureSpec.makeMeasureSpec(340, MeasureSpec.AT_MOST);
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }
}

File: module_views/src/main/java/com/monk/moduleviews/views/MaxHeightScrollView.kt
package com.monk.moduleviews.views

import android.app.Activity
import android.content.Context
import android.util.AttributeSet
import android.util.DisplayMetrics
import android.widget.ScrollView
import com.monk.commonutils.LogUtil

/**
 * @author monk
 * @date 2019-01-25
 */
class MaxHeightScrollView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : ScrollView(context, attrs) {
    private val tag = "MaxHeightScrollView"

    init {
        // 方式1
        val dm = context.resources.displayMetrics
        val sb = "heightPixels:${dm.heightPixels}\t\twidthPixels:${dm.widthPixels}\txDpi:${dm.xdpi}\tyDpi:${dm.ydpi}\tdensity:${dm.density}\t	densityDpi:${dm.densityDpi}	"
        LogUtil.i(tag, sb)
        LogUtil.v(tag, "toString():$dm")

        // 方式2
        val display = (context as Activity).windowManager.defaultDisplay
        val displayMetrics1 = DisplayMetrics()
        display.getMetrics(displayMetrics1)
        LogUtil.v(tag, "toString():$displayMetrics1")
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        var heightMeasureSpec = heightMeasureSpec
        heightMeasureSpec = MeasureSpec.makeMeasureSpec(340, MeasureSpec.AT_MOST)
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    }
}